### PR TITLE
First test with JSONFeed

### PR DESCRIPTION
**Pull request by uwearzt**
Todo:
- [x] Tests
- [x] check if valid
- [x] id
- [x] tags
- [ ] Refactoring

Please don't merge yet, just running CI, review remarks very welcome.

### CODE REVIEW

File: src/cobalt.rs [hunk 1]
```diff
@@ -281,6 +287,33 @@ fn create_rss(path: &str, dest: &Path, config: &Config, posts: &[Document]) -> R
...
+
+            let jsonfeed_string = jsonfeed::to_string(&feed).unwrap();
+            let jsonfeed_path = dest.join(path);
+            let mut jsonfeed_file = try!(File::create(&jsonfeed_path));
+            try!(jsonfeed_file.write_all(&jsonfeed_string.into_bytes()));
+
+            info!("Created jsonfeed file at {}", jsonfeed_path.display());
+            Ok(())
+        }
+        _ => Err(ErrorKind::ConfigFileMissingFields.into()),
```
Comment (epage) [inline src/cobalt.rs]: Personally, I try to have my indentation to reflect my business logic.

So while its "more code', I prefer
```rust
let name = config.name.ok_or(ErrorKind::ConfigFileMissingFields)?;
```

File: src/cobalt.rs [hunk 2]
```diff
@@ -281,6 +287,33 @@ fn create_rss(path: &str, dest: &Path, config: &Config, posts: &[Document]) -> R
...
+              description: Some(desc.to_string()),
+              ..Default::default()
+            };
+
+            let jsonfeed_string = jsonfeed::to_string(&feed).unwrap();
+            let jsonfeed_path = dest.join(path);
+            let mut jsonfeed_file = try!(File::create(&jsonfeed_path));
+            try!(jsonfeed_file.write_all(&jsonfeed_string.into_bytes()));
+
+            info!("Created jsonfeed file at {}", jsonfeed_path.display());
```
Comment (epage) [inline src/cobalt.rs]: Isn't `.display()` implicitly called?

File: src/document.rs [hunk 3]
```diff
@@ -295,6 +298,34 @@ impl Document {
...
+            .get("description")
+            .or_else(|| self.attributes.get("excerpt"))
+            .or_else(|| self.attributes.get("content"))
+            .and_then(|s| s.as_str())
+            .map(|s| s.to_owned());
+
+        let link = root_url.to_owned() + &self.path.replace("\\", "/");
+
+        Item {
+            id: "1".into(),
```
Comment (epage) [inline src/document.rs]: I assume every document needs a distinct ID and this is missing a TODO

File: src/document.rs [hunk 4]
```diff
@@ -295,6 +298,34 @@ impl Document {
...
+            url: Some(link.to_string()),
+            external_url: Some(link.to_string()),
+            title: description,
+            content: Content::Html("<p>content</p>".into()),
```
Comment (epage) [inline src/document.rs]: Missing a TODO?

File: src/document.rs [hunk 5]
```diff
@@ -295,6 +298,34 @@ impl Document {
...
+            summary: None,
+            image: None,
+            banner_image: None,
+            date_published: self.date.map(|date| date.to_rfc2822()),
+            date_modified: self.date.map(|date| date.to_rfc2822()),
+            author: None,
+            tags: Some(vec!["json".into(), "feed".into()]),
```
Comment (epage) [inline src/document.rs]: TODO?

**Comment by epage** [comment]
I'm not a web developer by trade.  How much traction has jsonfeed gotten?

File: src/cobalt.rs (hunk 2 above)
Comment (uwearzt) [inline src/cobalt.rs]: Nope:

info!("Created jsonfeed file at {}", jsonfeed_path);
    |                                                  ^^^^^^^^^^^^^ the trait `std::fmt::Display` is not implemented for `std::path::PathBuf`

This line was copied from the RSS feed creation :)

File: src/document.rs (hunk 3 above)
Comment (uwearzt) [inline src/document.rs]: Yes, I have added my TODOs to the git comment. Will move them into source in the next revision.

File: src/cobalt.rs (hunk 1 above)
Comment (uwearzt) [inline src/cobalt.rs]: That was also directly copied over from the rss generation. I will do a refactor step after implementing everything.

Comment (uwearzt) [inline src/cobalt.rs]: As far as I can see this will also needs some new functions called from rss and jsonfeed generation.

**Comment by uwearzt** [comment]
jsonfeed starts to get traction. I.e. https://newsblur.com has added support for jsonfeed. See also the podcast at https://github.com/cobalt-org/cobalt.rs/issues/248, especially Podcasters start to use jsonfeed.

File: src/cobalt.rs (hunk 1 above)
Comment (epage) [inline src/cobalt.rs]: Yeah, I've not had a chance to really get into that section of code yet.

File: src/document.rs [hunk 6]
```diff
@@ -308,21 +308,31 @@ impl Document {
...
+        let mut cat = Vec::new();
+        if let Some(categories) = self.attributes.get("categories") {
+            if let &Value::Array(ref categories_array) = categories {
+                for category in categories_array {
+                    if let &Value::Str(ref category_string) = category {
+                        cat.push(category_string.clone());
+                    }
+                }
+            }
+        }
```
Comment (epage) [inline src/document.rs]: Could you do something like
```rust
let cat: Vec<_> = self.attributes.get("categories").and_then(|v| v.as_array()).unwrap_or([]).iter().map(|v| v.to_string());
```

**Comment by epage** [comment]
Eh, anything else can be improved later.

**Pull request merged**